### PR TITLE
Synchronize `plotting arguments` and minor fixes

### PR DESCRIPTION
**Pull request by dh031200**
### Summary
🖼️🛠️📝

This pull request adds new configuration options and arguments to customize the output image annotations for different prediction modes of the YOLO model. It also updates the documentation, the default configuration file, and the argument names to reflect the changes and improve consistency. The affected files are `predict.md`, `cfg/__init__.py`, `results.py`, `plotting.py`, `predict.py`, `cfg.md`, and `default.yaml`.

> _`Predict` the doom of the world with new arguments_
> _Customize the output of your dark visions_
> _`Results` control the annotations of the end_
> _`Plot` the images of the final descent_

### Walkthrough
*  Add and update arguments for controlling the display of different annotations on the output images in the `predict` mode ([link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-31a8e95fab032c4aff065cd0faacd330d81e4b154c017a7827c9f85d7f8dea49L83-R88), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-31a8e95fab032c4aff065cd0faacd330d81e4b154c017a7827c9f85d7f8dea49L223-R240), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-a21b6f90c80bba2698f9a8e30b1a002078a425376e5eb2864f078e0a97d32a3bR145-R148), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-daf1714127b012666c9fed39d0e75184c60be83f0c6bd3f9863f27113b1a6c22L66-R71), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-9d692d4ad15d420a92105fcd432fb1d2221b71d5af055bea3bd2b70523ce6743L58-R64), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5R72-R78), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5L85-R115), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5R127-R133), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-d6b04521e58a824cca3a7cd6a53579b3ed5b91b1fa37d58a8544391a499c2aa2L30-R35), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-d47b2bd99d71bef47ccb82b27f8a102c04b98072070f94d0ba341fa3221f8b81L28-R35), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-1564811411239154c7f0b40eabff1f51a108b807b9ae746db09abbb70d5441bbL38-R42), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-ac57881f6ef056ee9e695ffb6f80c1a0356eed0b796005d8ce65ae35301bddb3L32-R39), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-ac57881f6ef056ee9e695ffb6f80c1a0356eed0b796005d8ce65ae35301bddb3L43-R58))
*  Replace the `line_width` argument with `line_thickness` for consistency and add a deprecation warning for the former in the `Results` and `Annotator` classes and the `plot_images` function ([link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-daf1714127b012666c9fed39d0e75184c60be83f0c6bd3f9863f27113b1a6c22R155-R157), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-bd2831d027341da2474db760605211ea00b347396b0d708a4343a2b709eb3675L151-R151), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5L161-R196), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5L179-R215), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5L200-R254), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5L221-R268), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-7e75b3f968b9868ef6f0a86148cf6c16a5224e7fdd45f5c2eb5e9bc0fc744756L50-R50), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-7e75b3f968b9868ef6f0a86148cf6c16a5224e7fdd45f5c2eb5e9bc0fc744756L67-R67), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-7e75b3f968b9868ef6f0a86148cf6c16a5224e7fdd45f5c2eb5e9bc0fc744756L345-R345))
*  Use the attributes of the `Results` class instead of the arguments of the `plot` function to control the display of different annotations on the output images and add the `keypoints` argument to the `plot` function ([link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5R207), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5L189-R226), [link](https://github.com/ultralytics/ultralytics/pull/2373/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5L232-R278))

#### 1. Added code so that arguments from the `YOLO` class can be passed to the `Results` class.
* PR for #2032 
* Added floating arguments missing from documentation.

#### 2. Synchronized different arguments and corrected document content.
Show_labels:
* show_labels:	https://docs.ultralytics.com/usage/cfg/?h=show_labels#predict  
* hide_labels:	https://docs.ultralytics.com/modes/predict/#arguments  

Line_thickness:
* line_width :	https://docs.ultralytics.com/modes/predict/#plotting-results	-> line_thickness
* line_thickness:	https://docs.ultralytics.com/modes/predict/#arguments

#### 3. For consistency, `show_keypoint` has been added to pose estimation.
`keypoints (bool): Whether to plot the keypoints.`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to result visualization and line width configuration.

### 📊 Key Changes
- Replaced various `hide_*` parameters with their `show_*` equivalents (`show_labels`, `show_conf`, `show_probs`, `show_boxes`, `show_masks`, `show_keypoints`).
- Changed `line_thickness` to `line_width` for bounding box specification, allowing dynamic scaling to image size.
- Updated documentation and examples to reflect the above changes.

### 🎯 Purpose & Impact
- 🤝 Improves user control over visualization details of object detection results, aiding interpretability.
- 📐 Allows more nuanced control over bounding box appearance, especially for different image resolutions.
- 📚 Ensures that documentation and tutorials stay current with the latest codebase for a seamless user experience.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@AyushExel what do you think of this?

**Comment by AyushExel** [comment]
@dh031200 thanks for your thoughful PR. Looks great! 
I actually prefer `line_width` as it is shorter to type. Why do you think like_thickness is better? Do you think we can just work with line_width instead? Everything else looks good!

**Comment by dh031200** [comment]
@AyushExel I like that better too 👍
I Replaced all `line_thickness` with `line_width`.

**Comment by Laughing-q** [comment]
@dh031200 Hey, thanks for the PR! some details I'd like to discuss with you:
- All the `show_*` in `Results.__init__` looks a bit verbose, we want to keep the constructor simple which only includes necessary attributes, i.e `boxes`, `masks`.
- Same to the `default.yaml` we want all the args minimal and necessary which I think `boxes` is enough to control whether to plot boxes by default. Here's  the logic:
    - for detection task, users usually want to visualize boxes which is fine by default.
    - for segmentation, we visualize both boxes and masks by default. Sometimes users feel the plotted boxes are noisy and do not want to plot boxes, but for most cases users would like to visualize masks(since it's segmentation task) which is currently fine since we plot masks by default, then we only need to provide an arg to control whether to plot boxes.(same to pose task).
- Point 2 is the default logic which we want all the things necessary and minimal, also we provide the custom way(Results.plot) to allow users to show/not show property(boxes, keypoints, masks) they want.
```python
results = model("image.jpg)
results[0].plot(boxes=True, masks=False)
results[0].plot(boxes=False, masks=True)
.
.
```

**Comment by AyushExel** [comment]
@Laughing-q agreed abuut the duplication part. It seems like now we have both boxes, and show_boxes to control the visibility. We only need one anf its better to stick with shorter argument

**Comment by dh031200** [comment]
@Laughing-q Basically, I worked based on the content of the document.
[Docs(predict)](https://docs.ultralytics.com/usage/cfg/#predict) shows that `predict()` can use `show_labels`, `show_conf`, `line_thickness` as arguments. However, existing code does not pass that argument when generating Results.

## ultralytics/yolo/v8/detect/predict.py
### Before
```python
results.append(Results(orig_img=orig_img, path=img_path, names=self.model.names, boxes=pred))
```

### After
```python
results.append(
                Results(orig_img=orig_img,
                        path=img_path,
                        names=self.model.names,
                        boxes=pred,
                        line_width=self.args.line_width,
                        show_conf=self.args.show_conf,
                        show_labels=self.args.show_labels,
                        show_boxes=self.args.show_boxes))
```

Also, `plot()` in [Docs(plotting-results)](https://docs.ultralytics.com/modes/predict/#plotting-results) also shows `line_width` and `boxes` as arguments. The arguments are different, but the role is the same. If either of them was passed, the `show_*` attribute had to be used to perform the function. And it was awkward to have only `line_width` and `boxes`, so I added another `plotting argument`.


#### ultralytics/yolo/engine/results.py
```python
def plot(): 
        ...
        show_conf = self.show_conf & conf
        show_labels = self.show_labels & labels
        pred_boxes, show_boxes = self.boxes, self.show_boxes & boxes
        pred_masks, show_masks = self.masks, self.show_masks & masks
        pred_probs, show_probs = self.probs, self.show_probs & probs
        pred_keypoints, show_keypoints = self.keypoints, self.show_keypoints & keypoints
```


Should we remove one of the usages of `predict()` and `plot()` to avoid duplication?

@AyushExel As mentioned above, the documents show both usages. And there is a use case in #2032 

I apologize in advance if I misunderstood your words or responded incorrectly due to my limited English proficiency. If that's the case, Would you mind explaining it to me once again?

**Comment by glenn-jocher** [comment]
@dh031200 hi there, 

Thank you for taking the time and effort to review the documentation and code of YOLOv8 and reporting this potential issue. 

After going through your message, it looks like there has been some redundancy in the implementation of `predict()` and `plot()`. While the documentation suggests the usage of `show_labels`, `show_conf`, `line_thickness`, `line_width`, `boxes` and `show_boxes`, the existing code only passes a subset of these arguments to the `Results` constructor, namely `boxes`, while the other arguments are taken care of through `show_*` attributes. 

As a solution, you suggested passing all of these arguments to the constructor, and using them uniformly across the codebase. While this is certainly a viable option, it might seem a bit verbose and redundant. 

Another approach could be to stick to using `show_*` attributes wherever possible, to keep the API clean and simple. This way, users can specify what they want to visualize by setting any of these attributes to `True` or `False`. Meanwhile, `line_width` can be used as it is shorter to type instead of `line_thickness`. In summary, we can avoid code redundancy by keeping the API simple and using attributes wherever possible. 

I hope this clarifies the issue for you. Let us know if you have any further questions or concerns. Thank you!

**Comment by Laughing-q** [comment]
@dh031200 actually we pass `show_conf` and `show_labels` here through `result.plot()`, so I don't think we need to pass these to Results right?
[https://github.com/ultralytics/ultralytics/blob/0ebd3f2959992be55f65cfc908bdf2fd3b618cef/ultralytics/yolo/engine/predictor.py#L148-L155](https://github.com/ultralytics/ultralytics/blob/0ebd3f2959992be55f65cfc908bdf2fd3b618cef/ultralytics/yolo/engine/predictor.py#L148-L155)
If we want to add `show_masks` and `show_keypoints` in our defaults then probably something like this would be better?
```python
            plot_args = dict(line_width=self.args.line_thickness,
                             boxes=self.args.boxes,
                             conf=self.args.show_conf,
                             labels=self.args.show_labels,
                             masks=self.args.show_masks,
                             keypoints=self.args.show_keypoints,
)
```

**Comment by Laughing-q** [comment]
@dh031200 I just came back from here [https://github.com/orgs/ultralytics/discussions/2032#discussioncomment-5786890](https://github.com/orgs/ultralytics/discussions/2032#discussioncomment-5786890) and saw your use case. But I'd like to stick to the `Before` usage, let me explain it to you:
- model.predict() api is designed to be out-of-box, and satisfy basic needs like save/show/plot image/videos and with some basic settings like the plot args in your example. The idea is like if you want to save or show plotted images then you don't need to use `result.plot()`, just pass `save=True` or `show=True`.
- result.plot() api is designed to users who want to do something based on the plotted image, and modify some args we do not include in default.yaml, i.e `font`, `font_size` and `pil`.

Basically model.predict() is for basic functionalities and result.plot() is for more customization. Also that's why I feel like `show_masks` and `show_keypoints` for default.yaml is also not necessary.

**Comment by dh031200** [comment]
@Laughing-q Yes, you are right.  So the code below works correctly.
```python
model.predict(verbose=False, show_labels=False, line_width=1, save=True)
```
In this way, when developing other software using yolov8, images must be loaded through functions such as `cv2.imread()`.

However, when trying to use an image directly from memory as shown below, you have no choice but to use `plot()`, but it does not work correctly.
```python
results = model.predict(verbose=False, show_labels=False, line_width=1)
annotated_image = results[0].plot()
#
# Some codes
#
cv2.imwrite('output.png', annotated_image)
```

I'll see if I can use `plot_args` to make the code more concise.

**Comment by Laughing-q** [comment]
@dh031200 yeah like I said above, for the custom use case that the basic settings in `model.predict()` do not satisfy users, I'd like to stick to the `Before` case.
```python
results = model.predict(verbose=False)
annotated_image = results[0].plot(labels=False, line_width=1)
#
# Some codes
#
cv2.imwrite('output.png', annotated_image)
```

**Comment by dh031200** [comment]
I think the issue has been clarified.
I'll close this PR and change 'line_thickness' to 'line_width'.

**Comment by Laughing-q** [comment]
@dh031200 Anyway, thanks for the effort and your thoughtful PR! :)
